sdk-rs/src/dlob.rs

```diff
@@ -110,6 +116,78 @@ impl DLOBClient {
             }
         });
 
+        RxStream(rx)
+    }
+
+    /// Subscribe to an orderbook via WebSocket.
+    pub async fn subscribe_ws(&self, market: MarketId) -> L2OrderbookStream {
+        // This unwrap should never panic
+        let ws_url = crate::utils::http_to_ws(&self.url).unwrap();
+        let (mut ws_stream, _) = connect_async(ws_url).await.unwrap();
+
+        // Setup channel for L2OrderbookStream
+        let (tx, rx) = channel::<Result<L2Orderbook, SdkError>>(16);
+
+        match market.kind {
+            MarketType::Perp => {
+                if let Some(perp_market_config) = perp_market_config_by_index(self.context, market.index) {
+                    let market_subscription_message = to_ws_json(perp_market_config);
+                    ws_stream.send(Message::Text(market_subscription_message)).await.map_err(|_| SdkError::SubscriptionFailure).unwrap();
+                }
+            },
+            MarketType::Spot => {
+                if let Some(spot_market_config) = spot_market_config_by_index(self.context, market.index) {
+                    let market_subscription_message = to_ws_json(spot_market_config);
+                    ws_stream.send(Message::Text(market_subscription_message)).await.map_err(|_| SdkError::SubscriptionFailure).unwrap();
+                }
+            }
+        }
+
+        let heartbeat_interval = TokioDuration::from_secs(5);
+        let mut last_heartbeat = Instant::now();
+        tokio::spawn(async move {
+            while let Some(message) = ws_stream.next().await {
+                if last_heartbeat.elapsed() > heartbeat_interval {
+                    eprintln!("Heartbeat missed!");
+                    let _ = ws_stream.close(None).await;
+                    break;              
+                }
+
+                match message {
+                    Ok(Message::Text(text)) => {
+                        let value: Value = serde_json::from_str(&text).unwrap_or_else(|_| json!({}));
+                        
+                        if value.get("channel").and_then(Value::as_str) == Some("heartbeat") {
+                            println!("Received heartbeat");
+                            last_heartbeat = Instant::now();
+                        } 
+                        else if let Some(channel) = value.get("channel").and_then(Value::as_str) {
+                            if channel.contains("orderbook") {
+                                let orderbook_data = value.get("data").and_then(Value::as_str).unwrap();
+                                match serde_json::from_str::<L2Orderbook>(&orderbook_data) {
```

Comment:
I like unwrap/pancking here since we relying on the format of the dlob-server and if its bad/changed then the user can't resolve it on their own so the service should panic